### PR TITLE
Implementing the option for Private DNS Zones in a separate Resource Group

### PR DESCRIPTION
**Pull request by cloudchristoph**
# Overview/Summary

Implements the Private DNS zones in a separate resource group.
Reason and behavior is described in Issue #164)

## This PR fixes/adds/changes/removes

1. Adds `privateDnsZones` as a separate module
2. Changes default switch and  within `hubNetworking` module

### Breaking Changes

1. Does not roll out Private DNS Zones for Private Link, if the parameter `parPrivateDNSZonesEnabled` is not explicitly set. (Param was true by default, and is now false by default)

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [ ] Associated it with relevant [ADO items](https://dev.azure.com/unifiedactiontracker/Solution%20Engineering/_backlogs/backlog/Azure%20Landing%20Zone%20Bicep/Backlog%20Bugs%20Feedback)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [ ] Performed testing and provided evidence.
- [ ] Updated relevant and associated documentation.

### CODE REVIEW

**Comment by ejhenry** [comment]
@cloudchristoph - would you be interested in updating this PR to align to the approach mentioned in #164 ? 

1. Split out Private DNS Zones to a separate module
2. Still call this module from Hub Networking and deploy to the default resource group
3. Investigate if we can easily parameterize the resource group to deploy the private DNS zones to inside of hub networking and default to the same resource group as it is today for parity reasons (as above) but also provide flexibility

**Comment by cloudchristoph** [comment]
> @cloudchristoph - would you be interested in updating this PR to align to the approach mentioned in #164 ?

Yes, sure!

> 1. Split out Private DNS Zones to a separate module

For this I assume we only have to change stuff in the original `hubNetworking` module - but I'll have a look.

> 2. Still call this module from Hub Networking and deploy to the default resource group

Yes.

> 3. Investigate if we can easily parameterize the resource group to deploy the private DNS zones to inside of hub networking and default to the same resource group as it is today for parity reasons (as above) but also provide flexibility
This will be possible. 

One question comes to mind: If I call the default `hubNetworking` module, I'll need a way to deploy Private DNS directly to the same RG (easy) and a way to deploy this to a separate group from either (1.) directly the hubNetworking module or (2.) by calling the privateDNS module separately from `hubNetworking`.

Let me think about that a little bit. ATM the separate call and a new parameter in the hubNetworking module to "disable P-DNS deployment" seems the best option.

**Comment by jtracey93** [comment]
> 

Thanks. We are also not against using changing the target scope of this template to `subscription` to be able to deploy to multiple resource groups etc. 👍

**Comment by ejhenry** [comment]
:+1: We can also leave 3 for a separate PR if you'd prefer

**Comment by ejhenry** [comment]
@cloudchristoph - Let us know if you're still working on the changes we discussed. If not, we will put in a new PR next week to close out the issue and will reference your contribution in the release notes.

**Comment by cloudchristoph** [comment]
Sorry for the late feedback. There was a lot of project business to do.

I have been able to implement all three items.

The `hubNetworking` module now calls the `privateDnsZones` module.
To avoid double maintenance of the zones in two modules, I have adjusted the parameter for the zones. If the parameter `parPrivateDnsZones` is not filled, all default zones are rolled out.

Furthermore, the zones are rolled out again by default, so there is no breaking change.

If you now additionally specify the parameter `parPrivateDNSZonesResourceGroup`, then the zones are rolled out in the separate resource group. No need to call the independent module separately.

There is still some documentation missing atm. But to give everyone deployment examples:

### Deploy Private DNS Zones to same Resource Group (the current default):

```bicep
targetScope = 'subscription'

param parRegion string = 'westeurope'
param parRgSuffix string = '01'

resource hubNetworkingRg 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: 'rg-core-network-weu-${parRgSuffix}'
  location: parRegion
}

module hubNetworking '../infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep' = {
  scope: hubNetworkingRg
  name: 'hubNetworking'
  params: {
    parRegion: parRegion
    parAzureFirewallEnabled: false
    parCompanyPrefix: 'alz'
    parDDoSEnabled: false
    parBastionEnabled: false
    parPrivateDNSZonesEnabled: true
    parVpnGatewayConfig: {}
    parExpressRouteGatewayConfig: {}
  }
}
```

### Deploy Private DNS Zones to different Resource Group:

```bicep
targetScope = 'subscription'

param parRegion string = 'westeurope'
param parRgSuffix string = '01'

resource hubNetworkingRg 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: 'rg-core-network-weu-${parRgSuffix}'
  location: parRegion
}

resource privateDnsZoneRg 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: 'rg-privatednszones-weu-${parRgSuffix}'
  location: parRegion
}

module hubNetworking '../infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep' = {
  scope: hubNetworkingRg
  name: 'hubNetworking'
  params: {
    parRegion: parRegion
    parAzureFirewallEnabled: false
    parCompanyPrefix: 'alz'
    parDDoSEnabled: false
    parBastionEnabled: false
    parPrivateDNSZonesEnabled: true
    parPrivateDNSZonesResourceGroup: privateDnsZoneRg.name
    parVpnGatewayConfig: {}
    parExpressRouteGatewayConfig: {}
  }
}
```


### Deploy Private DNS Zones as separate module:

```bicep
targetScope = 'subscription'

param parRegion string = 'westeurope'
param parRgSuffix string = '01'

resource hubNetworkingRg 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: 'rg-core-network-weu-${parRgSuffix}'
  location: parRegion
}

resource privateDnsZoneRg 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: 'rg-privatednszones-weu-${parRgSuffix}'
  location: parRegion
}

module hubNetworking '../infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep' = {
  scope: hubNetworkingRg
  name: 'hubNetworking'
  params: {
    parRegion: parRegion
    parAzureFirewallEnabled: false
    parCompanyPrefix: 'alz'
    parDDoSEnabled: false
    parBastionEnabled: false
    parPrivateDNSZonesEnabled: false
    parVpnGatewayConfig: {}
    parExpressRouteGatewayConfig: {}
  }
}

module privateDnsZones '../infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep' = {
  scope: privateDnsZoneRg
  name: 'privateDnsZones'
  params: {
    parRegion: parRegion
    parHubVirtualNetworkId: hubNetworking.outputs.outHubVirtualNetworkID
  }
}
```

Let me know what you guys think.

**Review by ejhenry — COMMENTED** [review]
Thanks @cloudchristoph looks good. A couple additions we'd like to include:

1. As you mentioned, a README doc for the privateDnsZones module and updates to the hubNetworking README.
2. Addition of the customerUsageAttribution module to privateDnsZones.bicep. You can use New-GUID to create a unique ID.
3. A screenshot or two of your test results in the testing evidence section of the PR.

File: infra-as-code/bicep/modules/privateDnsZones/README.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  -TemplateFile infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep `
+  -TemplateParameterFile infra-as-code/bicep/modules/privateDnsZones/mc-privateDnsZones.parameters.example.json
+  -ResourceGroupName 'Hub_PrivateDNS_POC'
+```
+## Example Output in Azure global regions
+
+![Example Deployment Output](media/privateDnsZonesExampleDeploymentOutput.png "Example Deployment Output in Azure global regions")
+
+## Example Output in Azure China regions
+![Example Deployment Output](media/mc-privateDnsZonesExampleDeploymentOutput.png "Example Deployment Output in Azure China")
```
Comment (cloudchristoph) [inline infra-as-code/bicep/modules/privateDnsZones/README.md:116]: Unfortunately, I don't have access to an Azure China region and can't show a sample deployment here.

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep [hunk 2]
```diff
@@ -0,0 +1,108 @@
+/*
+SUMMARY: Module to deploy the Private DNS Zones as per the Azure Landing Zone conceptual architecture 
+DESCRIPTION: The following components will deployed
+              Private DNS Zones - Details of all the Azure Private DNS zones can be found here --> https://docs.microsoft.com/en-us/azure/private-link/private-endpoint-dns#azure-services-dns-zone-configuration
+AUTHOR/S: aultt, jtracey93, cloudchristoph
+VERSION: 1.x.x
+*/
+@description('The Azure Region to deploy the resources into. Default: resourceGroup().location')
+param parRegion string = resourceGroup().location
```
Comment (ejhenry) [inline infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:9]: We have standardized on parLocation for specifying Azure regions (#174)

**Comment by ejhenry** [comment]
@cloudchristoph - the internal team is going to review the PR this Thursday. In the mean time can you resolve the conflicts with your branch?

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep (hunk 2 above)
Comment (cloudchristoph) [reply infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep]: Resolved with the last two commits.

**Comment by ejhenry** [comment]
@cloudchristoph - 👍 great work on this PR, very positive feedback from the team. Just a couple items to address:

1. For consistency with the other modules, the team prefers to stay with the existing format of using parPrivateDnsZones default values instead of using the varKnownDnsZones variable. Our thinking is that for now it is best to keep it as simple as possible for first time users to understand exactly what is being deployed and where the values are coming from.
2. Some inconsistent use of camel casing. E.g. - we want to use parPrivateDnsZonesEnabled instead of parPrivateDNSZonesEnable and . A lot of this is legacy from earlier work, but we're trying to clean it up with each PR
3. Remove 'copy' from the 'mc-privateDnsZones.parameters.example copy.json' file name

Thanks again for your contributions! Your efforts are very much appreciated!

cc: @jfaurskov

**Review by jtracey93 — CHANGES_REQUESTED** [review]
> making this a formal review 👍

@cloudchristoph - 👍 great work on this PR, very positive feedback from the team. Just a couple items to address:

For consistency with the other modules, the team prefers to stay with the existing format of using parPrivateDnsZones default values instead of using the varKnownDnsZones variable. Our thinking is that for now it is best to keep it as simple as possible for first time users to understand exactly what is being deployed and where the values are coming from.
Some inconsistent use of camel casing. E.g. - we want to use parPrivateDnsZonesEnabled instead of parPrivateDNSZonesEnable and . A lot of this is legacy from earlier work, but we're trying to clean it up with each PR
Remove 'copy' from the 'mc-privateDnsZones.parameters.example copy.json' file name
Thanks again for your contributions! Your efforts are very much appreciated!

cc: @jfaurskov Jan Faurskov FTE

File: infra-as-code/bicep/modules/privateDnsZones/README.md (hunk 1 above)
Comment (ejhenry) [reply infra-as-code/bicep/modules/privateDnsZones/README.md:115]: We can add the Azure China output example in a follow-on effort.

File: infra-as-code/bicep/modules/privateDnsZones/README.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+- Private DNS Zone Links - Links deployed zones with provided Hub Network
+
+## Parameters
+
+The module requires the following inputs:
+
+ | Parameter              | Type   | Default                                                                                                          | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | Requirement                              | Example                                                                                                                                                |
```
[the remark this answers is not in the document]
Comment (cloudchristoph) [reply infra-as-code/bicep/modules/privateDnsZones/README.md]: I've revised this Readme to make things clearer - and to get it under the Linter limit.
Let me know, if this works for you.

File: infra-as-code/bicep/modules/privateDnsZones/README.md (hunk 1 above)
Comment (cloudchristoph) [reply infra-as-code/bicep/modules/privateDnsZones/README.md]: done

**Review by ejhenry — APPROVED** [review]
LGTM

**Pull request merged**